[dashboard] Refactor What's New dialog

corneliusludmann opened the pull request:
#### That's how it looks like for an **active user** that has already **VSCode** as editor:

| Light Theme | Dark Theme |
| ------------- |-------------|
| ![image](https://user-images.githubusercontent.com/24960040/123450408-ff704000-d5dc-11eb-8ca5-61b7c7671870.png) | ![image](https://user-images.githubusercontent.com/24960040/123450670-452d0880-d5dd-11eb-8052-3de9975e58de.png) |


#### That's how it looks like for an **active user** that still has **Theia** as editor:

| Light Theme | Dark Theme |
| ------------- |-------------|
| ![image](https://user-images.githubusercontent.com/24960040/123451112-b2409e00-d5dd-11eb-8ae2-f2d581a7572d.png) | ![image](https://user-images.githubusercontent.com/24960040/123451032-9ccb7400-d5dd-11eb-9b96-57f48128eedc.png) |

After the click on “Close” or “X” the editor will be changed to VSCode. The user has to change this back in the preferences.


#### That's how it looks like for a user that has not been seen the previous “What's New” message because he/she has not logged in since then.

| Light Theme | Dark Theme |
| ------------- |-------------|
| ![Peek 2021-06-25 17-52](https://user-images.githubusercontent.com/24960040/123452032-4f9bd200-d5de-11eb-9c2f-9b0ec236f737.gif) | ![Peek 2021-06-25 17-52_2](https://user-images.githubusercontent.com/24960040/123452073-59bdd080-d5de-11eb-9787-74ae6746364c.gif) |




Fixes #4459

See also: #3956

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew-2021-06.tsx:
```diff
@@ -0,0 +1,43 @@
+            <div className="border-t border-b border-gray-200 dark:border-gray-800 -mx-6 px-6 py-4">
+                <p className="pb-2 text-gray-900 text-base font-medium">Exposing Ports</p>
+                <p className="pb-2 text-gray-500 text-sm">We've changed the default behavior of exposed ports to improve the security of your dev environments.</p>
+                <p className="pb-2 text-gray-500 text-sm">Exposing ports are not private by default. You can still change port visibility through the editor or even configure this with the visibility property in .gitpod.yml.</p>
```
**nitpick**: Small typo! Phew! 😅 

```suggestion
                <p className="pb-2 text-gray-500 text-sm">Exposing ports are now private by default. You can still change port visibility through the editor or even configure this with the visibility property in .gitpod.yml.</p>
```

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew.tsx:
```diff
@@ -0,0 +1,102 @@
+
+import { User } from "@gitpod/gitpod-protocol";
+import Modal from "../components/Modal";
+import { WhatsNewEntry202104 } from "./WhatsNew-2021-04";
+import { WhatsNewEntry202106 } from "./WhatsNew-2021-06";
```
**thought**: Interesting! 💭

gtsiolis reviewed (commented):
Thanks for adding this, @corneliusludmann! 🎁 

Left some early feedback in the [relevant discussion](https://gitpod.slack.com/archives/C01U6GF78AY/p1624553004125700?thread_ts=1624264095.102100&cid=C01U6GF78AY) (internal).

svenefftinge commented:
The main point for this release is to inform everyone who still is using Theia that support is going to be removed. I feel this message gets a little burried within the other information of the "what's new" dialog.

svenefftinge commented:
Could you add the deprecation and end-of-life message to the settings page as well?

corneliusludmann commented:
> Could you add the deprecation and end-of-life message to the settings page as well?

Done here: https://github.com/gitpod-io/gitpod/pull/4614

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew-2021-06.tsx:
```diff
@@ -0,0 +1,32 @@
+
+        return <>
+            <div className="border-t border-b border-gray-200 dark:border-gray-800 -mx-6 px-6 pt-6 pb-4">
+                <p className="pb-2 text-gray-900 dark:text-gray-100 text-base font-medium">Exposing Ports <span className="ml-2 px-3 py-1 uppercase bg-blue-50 text-blue-500 rounded-xl">Configuration Update</span></p>
```
**nitpick**: Color fix. 🎨 

```suggestion
                <p className="pb-2 text-gray-900 dark:text-gray-100 text-base font-medium">Exposing Ports <span className="ml-2 px-3 py-1 uppercase bg-blue-50 text-blue-500 dark:text-blue-100 dark:bg-blue-500 rounded-xl">Configuration Update</span></p>
```

gtsiolis commented on components/dashboard/src/components/CodeText.tsx:
```diff
@@ -0,0 +1,9 @@
+ */
+
+export default function CodeText(p: { children?: React.ReactNode }) {
+    return <span className="bg-gray-100 dark:bg-gray-800 px-1.5 py-1 rounded-md text-sm font-mono font-medium">{p.children}</span>;
```
**nitpick**: Minor color fix. 😇 

```suggestion
    return <span className="bg-gray-100 dark:bg-gray-700 px-1.5 py-1 rounded-md text-sm font-mono font-medium">{p.children}</span>;
```

gtsiolis reviewed (approved):
Almost there! 🏁

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew-2021-06.tsx:
```diff
@@ -0,0 +1,32 @@
+            </div>
+            { user.additionalData?.ideSettings?.defaultIde !== "code" && <>
+                <div className="border-b border-gray-200 dark:border-gray-800 -mx-6 px-6 pt-6 pb-4">
+                    <p className="pb-2 text-gray-900 dark:text-gray-100 text-base font-medium">New Editor <span className="ml-2 px-3 py-1 uppercase bg-orange-100 text-orange-700 dark:bg-orange-500 dark:text-orange-100 rounded-xl">Deprecation Warning</span></p>
```
**nitpick**: One more! 🌈 

```suggestion
                    <p className="pb-2 text-gray-900 dark:text-gray-100 text-base font-medium">New Editor <span className="ml-2 px-3 py-1 uppercase bg-orange-100 text-orange-700 dark:bg-orange-100 dark:text-orange-600 rounded-xl">Deprecation Warning</span></p>
```

gtsiolis replied:
**question**: Label font size could be `text-sm`, right?

gtsiolis reviewed (commented):
Left behind these two minor comments! 😇

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew-2021-06.tsx:
```diff
@@ -0,0 +1,33 @@
+        return <>
+            <div className="border-t border-b border-gray-200 dark:border-gray-800 -mx-6 px-6 pt-6 pb-4">
+                <p className="pb-2 text-gray-900 dark:text-gray-100 text-base font-medium">Exposing Ports <PillLabel>Configuration Update</PillLabel></p>
+                <p className="pb-2 text-gray-500 dark:text-gray-400 text-sm">We've changed the default behavior of exposed ports to improve the security of your dev environments.</p>
```
**question**: Would you consider the following change to be more accurate? Your call! 

```suggestion
                <p className="pb-2 text-gray-500 dark:text-gray-400 text-sm">We've changed the default behavior of exposed ports to improve the security of your workspaces.</p>
```

corneliusludmann replied:
Both sound good. I'm for keeping “dev environments”. Like a tad more.

gtsiolis replied:
DEAL

corneliusludmann commented:
Hey @gtsiolis! Thanks a lot for all your input! :heart: 

I've updated the PR as well as the PR description above with latest screenshots.

I noticed that the monospaced code texts overlap a bit with the text in the line above and below. If you have a quick fix for this handy, I can address this in this PR as well. Otherwise, I would vote for creating an issue and addressing this in a follow-up PR.

gtsiolis commented on components/dashboard/src/components/CodeText.tsx:
```diff
@@ -0,0 +1,9 @@
+ */
+
+export default function CodeText(p: { children?: React.ReactNode }) {
+    return <span className="bg-gray-100 dark:bg-gray-700 px-1.5 py-1 rounded-md text-sm font-mono font-medium">{p.children}</span>;
```
**nitpick**: This should fix the monospaced overlap issue.

```suggestion
    return <span className="bg-gray-100 dark:bg-gray-700 px-1.5 py-0.5 rounded-md text-sm font-mono font-medium">{p.children}</span>;
```

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew.tsx:
```diff
@@ -0,0 +1,100 @@
+                </>
+                :
+                <div className="flex justify-end mt-6 space-x-2">
+                    <button className="ml-2" onClick={internalClose}>Close</button>
```
**question**: What do you think of renaming this to Continue? Left margin is also not needed here now, right?

```suggestion
                    <button onClick={internalClose}>Continue</button>
```

Alternatively, we could use the secondary class here with the _Close_ label.

```suggestion
                    <button className="secondary" onClick={internalClose}>Close</button>
```

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew.tsx:
```diff
@@ -0,0 +1,100 @@
+                        <button className="ml-2 mt-0 secondary" onClick={internalClose}>Dismiss All</button>
+                        <div className="flex flex-col items-center">
+                            <button className="ml-2" onClick={next}>Next</button>
+                            <div className="text-xs italic mt-1 ml-2 -mb-2">({unseenEntries.length} left)</div>
```
**thought**: Thinking whether we could just drop this part. Maybe something to consider in https://github.com/gitpod-io/gitpod/issues/3956. 💭

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew.tsx:
```diff
@@ -0,0 +1,100 @@
+                        <button className="ml-2 mt-0 secondary" onClick={internalClose}>Dismiss All</button>
+                        <div className="flex flex-col items-center">
+                            <button className="ml-2" onClick={next}>Next</button>
+                            <div className="text-xs italic mt-1 ml-2 -mb-2">({unseenEntries.length} left)</div>
```
**suggestion**: Alternatively, we could left align this  on the same row with the button and read 1 of 2.

corneliusludmann replied:
Yes, I'm actually not that happy with the text under the “Next” button either. However, I personally don't like the schema “1 of 2” here because
1. there is no button back therefore the number of already seen pages seems not that important to me and
2. technically, I don't have the number of the total pages because the total number changes (decreases) at that moment I click on “Next“.

But we can address this in a follow-up PR if you like.

I would propose that change:

```diff
--- a/components/dashboard/src/whatsnew/WhatsNew.tsx
+++ b/components/dashboard/src/whatsnew/WhatsNew.tsx
@@ -76,12 +76,10 @@ export function WhatsNew(props: { onClose: () => void }) {
         {
             hasNext() ?
                 <>
-                    <div className="flex items-start justify-end mt-6 space-x-2">
-                        <button className="ml-2 mt-0 secondary" onClick={internalClose}>Dismiss All</button>
-                        <div className="flex flex-col items-center">
-                            <button className="ml-2" onClick={next}>Next</button>
-                            <div className="text-xs italic mt-1 ml-2 -mb-2">({unseenEntries.length} left)</div>
-                        </div>
+                    <div className="flex items-center justify-end mt-6 space-x-2">
+                        <div className="text-sm mr-auto italic">{unseenEntries.length} page{unseenEntries.length > 1 ? "s" : ""} left</div>
+                        <button className="ml-2 secondary" onClick={internalClose}>Dismiss All</button>
+                        <button className="ml-2" onClick={next}>Next</button>
                     </div>
                 </>
```

| Before        | After           |
| ------------- |-------------|
| ![image](https://user-images.githubusercontent.com/24960040/123469689-a8c23080-d5f3-11eb-8082-ce973e0b6dab.png)      |  ![image](https://user-images.githubusercontent.com/24960040/123470321-8977d300-d5f4-11eb-9c60-b7860a5780ab.png) |



![image](https://user-images.githubusercontent.com/24960040/123470370-998fb280-d5f4-11eb-96a1-85b915cdafe2.png)

What do you think?

gtsiolis replied:
LET'S DO IT. Visually it's much better. I would only rephrase _1 page left_ to _1 more update_ or _1 item left_ to move away from the page reference as this could be confusing. Your call! 📞

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew-2021-06.tsx:
```diff
@@ -0,0 +1,33 @@
+            { user.additionalData?.ideSettings?.defaultIde !== "code" && <>
+                <div className="border-b border-gray-200 dark:border-gray-800 -mx-6 px-6 pt-6 pb-4">
+                    <p className="pb-2 text-gray-900 dark:text-gray-100 text-base font-medium">New Editor <PillLabel type="warn">Deprecation Warning</PillLabel></p>
+                    <p className="pb-2 text-gray-500 dark:text-gray-400 text-sm">We're deprecating the Theia editor. You can still switch back to Theia for the next few weeks but the preference will be removed by the end of August 2021.</p>
```
**question**: Are we dropping the idea of having an action button to switch back to Theia from here? If no, a secondary button below the text and left aligned could suffice. Button could read _Switch back to Theia_. Alternatively, let's keep this as is.

corneliusludmann replied:
Actually, with the current implementation, it's not that easy to implement this. I would like to keep this as it for now.

gtsiolis replied:
LET'S DROP IT. 🏀

gtsiolis reviewed (approved):
Thanks for pushing this through the finish line @corneliusludmann!

Left some minor comments.

Shall we also remove the #3956 from the PR description and tackle this separately?

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew.tsx:
```diff
@@ -0,0 +1,100 @@
+            hasNext() ?
+                <>
+                    <div className="flex items-start justify-end mt-6 space-x-2">
+                        <button className="ml-2 mt-0 secondary" onClick={internalClose}>Dismiss All</button>
```
**praise**: Forgot to mention, thanks for adding this!

gtsiolis commented on components/dashboard/src/whatsnew/WhatsNew.tsx:
```diff
@@ -0,0 +1,100 @@
+                </>
+                :
+                <div className="flex justify-end mt-6 space-x-2">
+                    <button className="secondary" onClick={internalClose}>Continue</button>
```
**nitpick**: Last minute clarification for https://github.com/gitpod-io/gitpod/pull/4607#discussion_r658892570! If we rename this to _Continue_, using a primary button should be ok! 😇 

```suggestion
                    <button onClick={internalClose}>Continue</button>
```

corneliusludmann replied:
At first, I really liked “Continue” because it sounds more positive than “Close”. But then I thought about the difference between “Next” and “Continue”. For me, it sounds related/similar. When the button text changes from “Next” to “Continue” I probably ask myself “What's next? It is still not the end?” The text “Close” would it make more clear to me that we are done.

However, it's just a small nuance. For now, it's a primary button with “Continue” but can change it to a secondary button with “Close” anytime. Please let me know what you would prefer, @gtsiolis.

gtsiolis replied:
Primary (confim) button with a _Continue_ label sounds good for now! We'll get back to this in https://github.com/gitpod-io/gitpod/issues/3956. ✔️

corneliusludmann commented:
I've updated the PR. When we agree on the 2 left items we are ready to merge this on Monday. :rocket:

corneliusludmann commented on components/dashboard/src/whatsnew/WhatsNew-2021-06.tsx:
```diff
@@ -0,0 +1,33 @@
+        </>;
+    },
+    newsKey: 'June-2021',
+    maxUserCreationDate: '2021-06-30', // TODO
```
TODO: Find a suitable Date!

gtsiolis commented:
Thanks for the updates, @corneliusludmann! Left two comments for the unresolved discussions. 💬 

Friendly reminder to remove https://github.com/gitpod-io/gitpod/issues/3956 from the PR description as this contains suggestions out of the scope of this PR! Let me know if this you think this is not needed. 🎗️

corneliusludmann commented:
> Friendly reminder to remove #3956 from the PR description as this contains suggestions out of the scope of this PR! Let me know if this you think this is not needed.

Since it is not tagged with the `Fixes` keyword it should be fine to have it linked in the descriptions (it does not close the issue), right?

gtsiolis commented:
> ... it should be fine to have it linked in the descriptions ...

Absolutely! It's always helpful to have linked references like this. Thanks, @corneliusludmann! 🙏

corneliusludmann commented:
Thanks again, @gtsiolis!

- [x] Updated PR based on your comments.
- [x] No TODOs in the code left. :innocent:
- [x] Rebased on `main`.
- [x] Quick test of the preview env.

I think we are ready to merge. Any last-minute objections?

gtsiolis commented:
> Any last-minute objections?

LET'S DO THIS.
